// salesforce-source-gen/lib/commands/picklists.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const command_1 = require("@oclif/command");
const __1 = require("..");
class Picklists extends command_1.Command {
    async run() {
        const { flags: options } = this.parse(Picklists);
        __1.generatePicklistClasses(options);
    }
}
exports.default = Picklists;
Picklists.description = `Generate Apex constants from field picklists and standard value sets.
To filter only specific picklists, add to your package.json 'include'.
Only picklists defined in include will be generated.
Include can contain:
    I)   ObjectApiName - all custom fields of that object will be used
    II)  ObjectApiName.FieldApiName - custom field will be used
    III) StandardValueSetName - standard value set will be used
    IV)  GlobalValueSetName - global value set will be used
{
    "sourceGen": {
        "picklists": {
            "include": [
                "Account",
                "Building__c.Phase__c",
                "Industry"
            ]
        }
    }
}
`;
Picklists.flags = {
    help: command_1.flags.help({ char: "h" }),
    projectDir: command_1.flags.string({ description: "Default: current working directory." }),
    outputDir: command_1.flags.string({ description: "Default: default package directory." }),
    sourceApiVersion: command_1.flags.string({ description: "Default: from sfdx-project.json." }),
    ignorePicklists: command_1.flags.boolean({ description: "Ignore custom fields." }),
    ignoreStandardValueSets: command_1.flags.boolean({ description: "Ignore standard value sets." }),
    ignoreGlobalValueSets: command_1.flags.boolean({ description: "Ignore global value sets." }),
    customFieldPrefix: command_1.flags.string({
        description: "Prefix for classes generated from custom fields. Default: empty."
    }),
    customFieldSuffix: command_1.flags.string({
        description: "Suffix for classes generated from custom fields. Default: empty."
    }),
    customFieldInfix: command_1.flags.string({
        description: "String between sobject name and field name. Default: '_'."
    }),
    standardValueSetPrefix: command_1.flags.string({ description: "Prefix for classes generated from standard value sets. Default: empty." }),
    standardValueSetSuffix: command_1.flags.string({ description: "Suffix for classes generated from standard value sets. Default: empty." }),
    globalValueSetPrefix: command_1.flags.string({ description: "Suffix for classes generated from global value sets. Default: empty." }),
    globalValueSetSuffix: command_1.flags.string({ description: "Suffix for classes generated from global value sets. Default: empty." })
};
