// salesforce-source-gen/lib/commands/record-types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const command_1 = require("@oclif/command");
const __1 = require("..");
class RecordTypes extends command_1.Command {
    async run() {
        const { flags: options } = this.parse(RecordTypes);
        __1.generateRecordTypesClass(options);
    }
}
exports.default = RecordTypes;
RecordTypes.description = `Generate Apex constants from record types defined in your source.
You can use \`RecordTypes.ACCOUNT_AGENT_ID\`instead of \`Schema.SObjectType.Account.getRecordTypeInfosByDeveloperName().get('Agent');\`
or \`RecordTypes.ACCOUNT_AGENT\` to access RecordTypeInfo.`;
RecordTypes.flags = {
    help: command_1.flags.help({ char: "h" }),
    projectDir: command_1.flags.string({ description: "Default: current working directory" }),
    outputDir: command_1.flags.string({ description: "Default: default package directory from sfdx-project.json." }),
    outputClassName: command_1.flags.string({ description: "Default: 'RecordTypes'." }),
    sourceApiVersion: command_1.flags.string({ description: "Default: from sfdx-project.json." }),
    includeInactive: command_1.flags.boolean({ description: "Include inactive record types." }),
    ignoreTestClass: command_1.flags.boolean({ description: "Do not generate test class." })
};
