// salesforce-metadata/src/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var fs_extra_1 = __importDefault(require("fs-extra"));
var xml2js_1 = __importDefault(require("xml2js"));
var DEFAULT_BUILDER_OPTIONS = {
    xmldec: {
        version: "1.0",
        encoding: "UTF-8",
        standalone: undefined
    },
    renderOpts: {
        pretty: true,
        indent: "    ",
        newline: "\n"
    }
};
function readFile(file) {
    return fs_extra_1.default.readFile(file);
}
// noinspection JSUnusedGlobalSymbols
function readMetadataXML(file) {
    var parser = new xml2js_1.default.Parser({});
    return readFile(file).then(function (data) { return parser.parseStringPromise(data.toString()); });
}
exports.readMetadataXML = readMetadataXML;
// noinspection JSUnusedGlobalSymbols
function writeMetadataXML(file, metadata, options) {
    var builder = new xml2js_1.default.Builder(options || DEFAULT_BUILDER_OPTIONS);
    return fs_extra_1.default.writeFile(file, builder.buildObject(metadata) + "\n");
}
exports.writeMetadataXML = writeMetadataXML;
__export(require("./metadata-types"));
var describe_metadata_1 = require("./describe/describe-metadata");
exports.describeMetadata = describe_metadata_1.describeMetadata;
__export(require("./find-metadata"));
